presto-native-execution/presto_cpp/main/operators/PartitionAndSerialize.cpp

```diff
@@ -68,19 +73,135 @@ class PartitionAndSerializeOperator : public Operator {
+    int32_t maxOutputSize = numInput;
+    if (replicateNullsAndAny_) {
+      const int32_t nullCount = nullRows_.countSelected();
+      // TODO: If numPartitions_ and nullCount is large, we need to batch out
+      // getOutput logic
+      maxOutputSize = numInput + (nullCount * numPartitions_) + numPartitions_;
+    }
+    return maxOutputSize;
+  }
+
+  // If replicateNullsAndAny_ is true, we need to duplicate null values across
+  // all partition ids. For efficiency, we calculate how many nulls needs to be
+  // added to each partition id.
+  // Returns the total elements in partitionVector
+  vector_size_t fillDataWithNull(
+      FlatVector<StringView>& dataVector,
+      FlatVector<int32_t> partitionVector) {
+    auto numInput = input_->size();
+    auto insertInto = numInput;
+
+    for (auto forPartition = 0; forPartition < numPartitions_; ++forPartition) {
+      for (auto copyFromIdx = 0; copyFromIdx < numInput; ++copyFromIdx) {
+        // Copy only if copyFromIdx has null partition key and it is not already
+        // assigned to forPartition.
+        // If Partition key is null, copy the serialised data into target index
+        if (nullRows_.isValid(copyFromIdx) &&
+            partitions_[copyFromIdx] != forPartition) {
+          dataVector.setNoCopy(insertInto, dataVector.valueAt(copyFromIdx));
+          partitionVector.set(insertInto, forPartition);
+          insertInto++;
+        }
+      }
+    }
+    return insertInto;
+  }
+
+  // Identify the first non-empty element and replicate on all empty partitions
+  vector_size_t replicateAnyInEmptyPartitions(
+      FlatVector<int32_t>& partitionsVector,
+      FlatVector<StringView>& dataVector,
+      const vector_size_t start) {
+    auto rawPartitions = partitionsVector.mutableRawValues();
+    const auto numInput = input_->size();
+    std::set<int32_t> nonEmptyPartitions(
+        rawPartitions, rawPartitions + numInput);
+    StringView firstNonEmpty;
+    for (int idx = 0; idx < numInput; ++idx) {
+      if (!nullRows_.isValid(idx)) {
+        firstNonEmpty = dataVector.valueAt(idx);
+        break;
+      }
+    }
+
+    int32_t finalOutputSize = start;
+    for (auto idx = 0; idx < numPartitions_; ++idx) {
+      if (nonEmptyPartitions.find(idx) == nonEmptyPartitions.end()) {
+        rawPartitions[finalOutputSize] = idx;
+        dataVector.setNoCopy(finalOutputSize, firstNonEmpty);
+        finalOutputSize++;
+      }
+    }
+    return finalOutputSize;
+  }
+
+  // A partition key is considered null only if its recursive children all are
+  // nulls. collectNullRows will flatten the key columns and store the null bits
+  // in a SelectivityVector nullRows_
+  void collectNullRows() {
+    auto size = input_->size();
+    // if replicateNullsAndAny_ is false, we don't need to calcualte nullRows_
+    if (!replicateNullsAndAny_) {
+      nullRows_.resize(size);
+      nullRows_.clearAll();
+      return;
+    }
+    rows_.resize(size);
```

Comment:
What's the use of rows_?

Reply:
It is used internally by velox::vector::DecodedVector to identify which rows to `.decode()`, for our use-case we need to decode all rows, hence we use it with `rows_.setAll();`.
I found a better function for `.decode()` which does not require rows_ to be passed in. Updated to use that function instead.